domain/network/service/interface.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/juju/core/network"
+)
+
+// State describes retrieval and persistence methods needed for the network
+// domain service.
+type State interface {
+	// Space (sub-domain) methods
+
+	// AddSpace creates and returns a new space.
+	AddSpace(ctx context.Context, uuid string, name string, providerID network.Id, subnetIDs []string) error
+	// GetSpace returns the space by UUID.
+	GetSpace(ctx context.Context, uuid string) (*network.SpaceInfo, error)
+	// GetSpaceByName returns the space by name.
+	GetSpaceByName(ctx context.Context, name string) (*network.SpaceInfo, error)
+	// GetAllSpaces returns all spaces for the model.
+	GetAllSpaces(ctx context.Context) (network.SpaceInfos, error)
+	// UpdateSpace updates the space identified by the passed uuid.
+	UpdateSpace(ctx context.Context, uuid string, name string) error
+	// DeleteSpace deletes the space identified by the passed uuid.
+	DeleteSpace(ctx context.Context, uuid string) error
+
+	// Subnet (sub-domain) methods
+
+	// AddSubnet creates a subnet.
+	AddSubnet(ctx context.Context, uuid string, subnet network.SubnetInfo) error
+	// GetAllSubnets returns all known subnets in the model.
+	GetAllSubnets(ctx context.Context) (network.SubnetInfos, error)
+	// GetSubnet returns the subnet by UUID.
+	GetSubnet(ctx context.Context, uuid string) (*network.SubnetInfo, error)
+	// GetSubnetsByCIDR returns the subnets by CIDR.
+	// Deprecated, this method should be removed when we re-work the API
+	// for moving subnets.
+	GetSubnetsByCIDR(ctx context.Context, cidrs ...string) (network.SubnetInfos, error)
+	// UpdateSubnet updates the subnet identified by the passed uuid.
+	UpdateSubnet(ctx context.Context, uuid string, spaceID string) error
+	// DeleteSubnet deletes the subnet identified by the passed uuid.
+	DeleteSubnet(ctx context.Context, uuid string) error
+	// UpsertSubnets updates or adds each one of the provided subnets in one
+	// transaction.
+	UpsertSubnets(ctx context.Context, subnets []network.SubnetInfo) error
+}
```

Comment:
Can we create a `SpaceState` and a `SubneState` and then `State` is just a composition of those two?

Reply:
yeah I like it! see fd1e0f0b87